controllers/cruisecontroloperation_controller.go

```diff
@@ -0,0 +1,508 @@
+	client.Client
+	DirectClient client.Reader
+	Scheme       *runtime.Scheme
+	Scaler       scale.CruiseControlScaler
+}
+
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations/finalizers,verbs=update
+
+//nolint:gocyclo
+func (r *CruiseControlOperationReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+	log.V(1).Info("reconciling CruiseControlOperation custom resources")
+
+	ccOperationListClusterWide := banzaiv1alpha1.CruiseControlOperationList{}
+	err := r.DirectClient.List(ctx, &ccOperationListClusterWide, client.ListOption(client.InNamespace(request.Namespace)))
+	if err != nil {
+		return requeueWithError(log, err.Error(), err)
+	}
+
+	// Selecting the reconciling CruiseControlOperation
+	var currentCCOperation *banzaiv1alpha1.CruiseControlOperation
+	for i := range ccOperationListClusterWide.Items {
+		ccOperation := &ccOperationListClusterWide.Items[i]
+		if ccOperation.GetName() == request.Name && ccOperation.GetNamespace() == request.Namespace {
+			currentCCOperation = ccOperation
+			break
+		}
+	}
+
+	if currentCCOperation == nil || r.isCCTaskTest(currentCCOperation) {
+		return reconciled()
+	}
+
+	// When the task is done we can remove the finalizer instantly thus we can return fast here.
+	if isFinalizerNeeded(currentCCOperation) && currentCCOperation.IsDone() {
+		controllerutil.RemoveFinalizer(currentCCOperation, ccOperationFinalizerGroup)
+		if err := r.Update(ctx, currentCCOperation); err != nil {
+			return requeueWithError(log, "error happened when removing finalizer", err)
+		}
+		return reconciled()
+	}
+
+	// CruiseControlOperation has invalid operation type, reconciled
+	if !currentCCOperation.IsCurrentTaskOperationValid() {
+		log.Error(errors.New("Koperator does not support this operation"), "operation", currentCCOperation.CurrentTaskOperation())
+		return reconciled()
+	}
+
+	kafkaClusterRef, err := kafkaClusterReference(currentCCOperation)
+	if err != nil {
+		return requeueWithError(log, "couldn't get kafka cluster reference", err)
+	}
+
+	kafkaCluster := &banzaiv1beta1.KafkaCluster{}
+	err = r.Get(ctx, kafkaClusterRef, kafkaCluster)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			if !currentCCOperation.ObjectMeta.DeletionTimestamp.IsZero() {
+				log.Info("kafka cluster has been removed, now removing finalizer from CruiseControlOperation")
+				controllerutil.RemoveFinalizer(currentCCOperation, ccOperationFinalizerGroup)
+				if err := r.Update(ctx, currentCCOperation); err != nil {
+					return requeueWithError(log, "failed to remove finalizer from CruiseControlOperation", err)
+				}
+			}
+			return reconciled()
```

Comment:
Why do we just ignore the not found error here and return reconciled()?

Reply:
The referenced kafkaCluster CR has been deleted thus there is no Kafka cluster anymore we can remove the finalizer from the current CruiseControlOperation when it is under deletion and we are done with the current CCOperation. There is no execution needed. This is not an error. is my explanation reasonable?